test/functional/shared.js

```diff
@@ -72,76 +135,58 @@ function makeCleanupFn(client) {
         return reject(err || e);
       }
     });
-  };
-}
+  }
+
+  function lambda() {
+    if (!client) {
+      client = this.configuration.newClient(connectionString);
+    }
+    return client
```

Comment:
I don't really know how much of a goal this is for us, but this does force users of this method to always treat the return type of this method as a `Promise` (as evidenced by below tests I believed). Did you try to write it in a way that would accommodate users just using callbacks? I think it means this would be more cumbersome to use if I had wanted to write this:

```
it('should do something', function(done) {
  const blah = setupSomething();

  withClient(client => {
    // do something with client
    done();
  });
});
```
now I think I would have to do:
```
it('should do something', function(done) {
  const blah = setupSomething();

  withClient(client => {
    // do something with client
    done();
  }).then(done);
});
```

Reply:
I'm not sure there's any way to support the first kind of test exactly how it's written; `withClient` either needs to get a client to use as an argument, or be bound to the test function's `this` so it can make one.

The easiest option would be just doing the setup once you already have the client:

```
it('should do something', withClient((client, done) => {
  const blah = setupSomething();
  // do something with client
  done();
});
```

Or, if the setup has to happen first, you could return `withClient` inside the test, though this is a little more unwieldy. 

```
it('should do something', function() {
  const blah = setupSomething();
  return withClient.bind(this)(client, done) => {
    // do something with client
    done();
}

Technically the second option is using promises under the hood, but there aren't any promises in the individual test code.
```

Reply:
SGTM